File: code_samples/section11/problems/problem11_1/problem11_1.cs
static int[] BfsShortestPaths(int n, List<List<int>> adj, int s) {
    // Computes shortest path distances from a single source in an UNWEIGHTED graph
    // using Breadth-First Search (BFS).
    //
    // Parameters:
    //   n   : number of vertices (assumed labeled 0..n-1)
    //   adj : adjacency list; adj[u] contains neighbors of u
    //   s   : source vertex index
    //
    // Returns:
    //   dist[v] = minimum number of edges from s to v
    //   dist[v] = -1 if v is unreachable from s
    //
    // Notes:
    //   - BFS is correct for unweighted graphs because it explores vertices
    //     in increasing distance layers from the source.
    //   - -1 is used as a sentinel meaning "unvisited/unreachable".
    var dist = new int[n];
    Array.Fill(dist, -1);           // initialize all distances to -1 (unvisited/unreachable)

    var q = new Queue<int>();       // FIFO queue for BFS frontier

    dist[s] = 0;                    // distance from source to itself is 0
    q.Enqueue(s);                   // start BFS from the source

    while (q.Count > 0) {
        int u = q.Dequeue();        // dequeue next vertex to expand

        // Explore all neighbors of u.
        foreach (var v in adj[u]) {
            // If v hasn't been visited yet, the first time we discover it is
            // via the shortest path (in an unweighted graph).
            if (dist[v] == -1) {
                dist[v] = dist[u] + 1; // neighbor is one edge farther than u
                q.Enqueue(v);          // enqueue v for later expansion
            }
        }
    }

    return dist;
}

// ===== Helper functions =====

List<List<int>> MakeGraph(int n)
{
    // Allocates an empty adjacency list for a graph with n vertices.
    //
    // Returns:
    //   g : List<List<int>> where each g[i] starts as an empty neighbor list.
    //
    // Note:
    //   - This creates the container; edges are added separately via AddEdge().
    var g = new List<List<int>>(n);

    // Pre-populate with n empty lists so g[i] is valid for all 0..n-1.
    for (int i = 0; i < n; i++)
        g.Add([]); // target-typed "new List<int>()" (C# collection expression / modern syntax)
    return g;
}

void AddEdge(List<List<int>> adj, int u, int v)
{
    // Adds an UNDIRECTED edge between u and v.
    //
    // Parameters:
    //   adj : adjacency list
    //   u,v : endpoints
    //
    // Behavior:
    //   - Adds v to adj[u] and u to adj[v].
    //   - No bounds checking is performed on u and v.
    adj[u].Add(v);
    adj[v].Add(u);
}

void PrintTest(string name, int[] dist, int[] expected)
{
    // Simple "print and compare" helper for tests.
    //
    // Parameters:
    //   name     : label for the test case
    //   dist     : computed distances to print
    //   expected : expected distances to print
    //
    // Note:
    //   - This does not assert correctness; it just prints both arrays for visual comparison.
    Console.WriteLine(name);
    Console.WriteLine("Distances: " + string.Join(", ", dist));
    Console.WriteLine("Expected : " + string.Join(", ", expected));
    Console.WriteLine();
}

// ===== Test 1: Connected graph =====

{
    Console.WriteLine("=== Test 1: Connected graph ===");

    // Graph (undirected):
    // 0 - 1 - 2 - 4
    //     |
    //     3
    //
    // From source 0, expected shortest distances:
    // [0, 1, 2, 2, 3]
    int n = 5;
    var adj = MakeGraph(n);

    // Add undirected edges to match the diagram.
    AddEdge(adj, 0, 1);
    AddEdge(adj, 1, 2);
    AddEdge(adj, 1, 3);
    AddEdge(adj, 2, 4);

    // Run BFS shortest paths from node 0.
    int[] dist = BfsShortestPaths(n, adj, 0);

    // Expected distance array for visual verification.
    int[] expected = [0, 1, 2, 2, 3];

    PrintTest("Connected graph", dist, expected);
}

// ===== Test 2: Disconnected graph =====

{
    Console.WriteLine("=== Test 2: Disconnected graph ===");

    // Graph (undirected):
    // 0 - 1    2    3 - 4
    //
    // From source 0:
    //   - nodes 0 and 1 are reachable
    //   - nodes 2, 3, 4 are unreachable => -1
    int n = 5;
    var adj = MakeGraph(n);

    // Two separate components: {0,1} and {3,4}; node 2 is isolated.
    AddEdge(adj, 0, 1);
    AddEdge(adj, 3, 4);

    int[] dist = BfsShortestPaths(n, adj, 0);
    int[] expected = [0, 1, -1, -1, -1];

    PrintTest("Disconnected graph", dist, expected);
}

// ===== Test 3: Single node graph =====

{
    Console.WriteLine("=== Test 3: Single node graph ===");

    // Graph with a single vertex 0 and no edges.
    // From source 0: dist should be [0].
    int n = 1;

    // Adjacency list has one empty neighbor list (no edges to add).
    var adj = MakeGraph(n);

    int[] dist = BfsShortestPaths(n, adj, 0);
    int[] expected = [0];

    PrintTest("Single node graph", dist, expected);
}

File: code_samples/section11/problems/problem11_1/problem11_1.js
// adj: adjacency list as array of arrays, nodes 0..n-1
function bfsShortestPaths(n, adj, s) {
    // Computes shortest path distances from a single source in an UNWEIGHTED graph
    // using Breadth-First Search (BFS).
    //
    // Parameters:
    //   n   : number of nodes (assumed labeled 0..n-1)
    //   adj : adjacency list; adj[u] is an array of neighbors of node u
    //   s   : source node index
    //
    // Returns:
    //   dist : array of length n where:
    //          - dist[v] is the minimum number of edges from s to v
    //          - dist[v] is -1 if v is unreachable from s
    //
    // Implementation detail:
    //   - Uses a plain array as a queue plus a moving head index so we avoid O(n)
    //     shift() operations.
    const dist = new Array(n).fill(-1); // -1 marks "unvisited/unreachable"
    const q = [];                       // queue storage
    let head = 0;                       // index of next element to dequeue

    // Initialize BFS from the source.
    dist[s] = 0;
    q.push(s);

    // Process the queue in FIFO order.
    while (head < q.length) {
        const u = q[head++];            // dequeue next node to expand

        // Explore all neighbors of u.
        for (const v of adj[u]) {
            // If v has not yet been discovered, record its distance and enqueue it.
            // In an unweighted graph, first discovery gives the shortest distance.
            if (dist[v] === -1) {
                dist[v] = dist[u] + 1;
                q.push(v);
            }
        }
    }

    return dist;
}

function bfsShortestPaths(n, adj, s) {
    // NOTE: This is a duplicate definition of bfsShortestPaths.
    //
    // In JavaScript, the second function with the same name overrides/replaces
    // the earlier one in the same scope. That means *this* version is the one
    // that will actually be used by the test harness below.
    //
    // The logic is identical to the first definition: BFS for unweighted shortest paths.
    const dist = new Array(n).fill(-1);
    const q = [];
    let head = 0;

    dist[s] = 0;
    q.push(s);

    while (head < q.length) {
        const u = q[head++];
        for (const v of adj[u]) {
            if (dist[v] === -1) {
                dist[v] = dist[u] + 1;
                q.push(v);
            }
        }
    }
    return dist;
}

// ====================================
// Helpers
// ====================================

function makeGraph(n) {
    // Allocates an empty adjacency list for a graph with n nodes.
    //
    // Returns:
    //   adj : array of length n, where each adj[i] is an empty array of neighbors.
    return Array.from({ length: n }, () => []);
}

function addEdge(adj, u, v) {
    // Adds an UNDIRECTED edge between u and v in an adjacency list.
    //
    // Parameters:
    //   adj : adjacency list
    //   u,v : endpoints
    //
    // Behavior:
    //   - Adds v to u's neighbor list and u to v's neighbor list.
    //   - No bounds checking is performed.
    adj[u].push(v);
    adj[v].push(u);
}

function runTest(name, n, adj, src, expected) {
    // Runs bfsShortestPaths on a test graph and prints:
    //   - test name
    //   - source node
    //   - computed distances
    //   - expected distances
    //
    // This is a "print and compare" harness (no assertions).
    console.log(name);
    console.log(`Source = ${src}`);

    const dist = bfsShortestPaths(n, adj, src);

    console.log("Distances:", JSON.stringify(dist));
    console.log("Expected :", JSON.stringify(expected));
    console.log();
}

// ====================================
// Test Cases
// ====================================

function testConnected() {
    // Graph:
    // 0 - 1 - 2 - 4
    //     |
    //     3
    //
    // From 0 → distances: [0,1,2,2,3]
    const n = 5;
    const adj = makeGraph(n);

    // Add undirected edges to match the diagram.
    addEdge(adj, 0, 1);
    addEdge(adj, 1, 2);
    addEdge(adj, 1, 3);
    addEdge(adj, 2, 4);

    runTest("=== Test 1: Connected Graph ===", n, adj, 0, [0, 1, 2, 2, 3]);
}

function testDisconnected() {
    // Graph:
    // 0 - 1    2    3 - 4
    //
    // From 0 → distances: [0,1,-1,-1,-1]
    const n = 5;
    const adj = makeGraph(n);

    // Two separate components {0,1} and {3,4}; node 2 is isolated.
    addEdge(adj, 0, 1);
    addEdge(adj, 3, 4);

    runTest("=== Test 2: Disconnected Graph ===", n, adj, 0, [0, 1, -1, -1, -1]);
}

function testSingleNode() {
    // Graph:
    // n = 1, no edges
    //
    // From 0 → [0]
    const n = 1;
    const adj = makeGraph(n); // adjacency list with one empty neighbor list

    runTest("=== Test 3: Single Node Graph ===", n, adj, 0, [0]);
}

// ====================================
// Run Tests
// ====================================

// Execute tests in order.
testConnected();
testDisconnected();
testSingleNode();
